### ISSUE TITLE
Implement vote progress bar for participants

### ISSUE THREAD

**Issue by patcon** [pol-is/issues#29]
Participants currently have no way to see how close they are to "done".

As I understand it, this psychologically makes people less likely to continue, as they have no sense of when "success" (ie. completion) might be.

Research Paper: [The importance of percent-done progress indicators for computer-human interfaces](https://dl.acm.org/citation.cfm?id=317459)

If the thought is that numbers might be daunting at first, and discourage people from starting, then perhaps it could only show when people have proven themselves engaged. So if after 10 votes, the conversation starts showing a progress bar, then I think people would be much more likely to carry on until the end.

Thoughts?

**Comment by colinmegill** [comment]
A/B test the only way to know... if we implement it on all conversations we
could see whether there was an uptick or downtick relative to comprable
convos.

**Comment by mbjorkegren** [comment]
A/B test is a great idea. 
I had a look at the study, but it focuses on the kind of progress bars that are shown when the computer is doing the work.

**Comment by patcon** [comment]
@mbjorkegren Ah, I took the primary resource from a more recent (and perhaps more convince) article I was reading:
https://blog.kissmetrics.com/the-progress-bar/

PS - interactive vs non-interactive doesn't strike me as apples-to-oranges. I've heard similar stories about "next train" countdown signs *increasing* satisfaction with waiting on subway platform. As I recall, humans will post-facto express more satisfaction with a known wait, than a short unknown wait. <sup>[citation needed]</sup> This seems relevant to us here :)

**Comment by mbjorkegren** [comment]
Here are a few pol.is-specific aspects that I think we'd need to think about:
1. As people add comments, your progress will potentially go backwards.
2. How do we define "done"? Do you need to vote on every single comment, including comments flagged as low quality? Or if there are only 5 comments in the conversation, are you really done if you've voted on those 5?

**Comment by patcon** [comment]
In my thinking, it was a countdown until the "enter your email" field, at which point the tool can help you remember to visit again. Until then, you have to remember to come back to finish, if you didn't have time during this visit. Email subscription seems like the current "winning" situation -- you get access to a means of engagement that isn't accessible via any other path.

1. I'd be alright with the progress bar going backward (presumably refreshing with the graph itself and comment itself? Perhaps before.)
2. re: when are we done? I would think that this is something we'll need to define anyhow, regardless of this feature request. (or else, when do we ask people about subscribing?) But to give my peanut gallery thoughts: I would say its a countdown until the "there are not more actions to perform" message. as in, low quality questions are included. I'd say it's up to moderator/crowd to remove those, but until then, a progress bar would include them. re: low number (5) questions. I think it's a countdown until "no more actions to throw at user", so the copy should be careful not to let people believe they're "done" this conversation -- just that we have nothing left to offer them right now. I feel like we're in the same boat as status bars that encourage people to complete their profiles during sign-up -- I don't feel users who are chasing that "completion" expect that their journey on the site is over. Though I suppose it's our job to be clear on that with the lingo :)

**Comment by patcon** [comment]
Maybe this is making this too large a UX change, but perhaps progress indicator is the wrong UI visual representation. Perhaps "stacks of cards" is a better one. So there could be a stack of cards on the right (with a number on it, showing how many comments there are unseen), and a stack on the left (with a number, showing how many voted on), and voting on comments increments and decrements appropriately. Adding some motion to the incoming and outgoing comments (instead of the fade), might further the analogy of comments to cards, moving from one pile to another.

But, this is perhaps out of scope.

**Comment by colinmegill** [comment]
How about red yellow green zones that stay constant on the bar. If there
are only 5 comments you cannot get out of the red so you have to come back.
Green is over 50 or something. It also shows the max, which increases but
doesn't ding your absolute progress, just looks like red and yellow shrink.

**Comment by patcon** [comment]
I'm not sure I follow on the "5 comments" special case, for which you both seem to be on the same page: Why is a small number of comments a special case where we can't tell people they're "done" for now?

It sounds like you're both thinking on longer-term gamification (like an overall per-user conversation score system), and I'm just trying to figure out how to encourage people to finish what's available to them _right now_ without aborting prematurely (short-term). Is this two separate issues, perhaps?

**Comment by colinmegill** [comment]
There would be a bit of copy with this. Perhaps "this conversation is just getting started, stop back again when there are more comments to vote on." We don't want people to think they're done when there are only 5 comments, they aren't. Of course, we also aren't doing a good job right now of letting people know that there will be more in the future other than subscribe. We should have better language around that.

![image](https://cloud.githubusercontent.com/assets/1770265/21158100/21518ebc-c139-11e6-8e2a-f212d9be5a7a.png)

**Comment by patcon** [comment]
Ah I understand what you mean now. I can see this being one way.

Having said that, I will admit that I don't feel that the colours add much -- and i'm unsure why you would choose colours over a straight-forward completeness meter. I have this sense that colours should mean something -- red means bad. But there's honestly no "losing"ness in only completing 5 comments. There are no points or anything to justify that scheme, right? I mean, it doesn't hurt, but it's kinda unnecessary noise in the UI, as far as I can imagine anyone caring :)

**Comment by patcon** [comment]
Yay. Someone already pirated my favourite book:
http://dl.finebook.ir/book/6d/10869.pdf

> 3.12. Misusing or Overusing Color

**Comment by colinmegill** [comment]
Iterate on the design and post a better one!

**Comment by colinmegill** [comment]
Definitely disagree that 5 votes is enough. If everyone did that, our output would be next to meaningless because of sparsity. 

Ditching the colors is certainly desirable because it overloads the agree and disagree colors. But if the idea, as to your first point, is to compel the user past 24-27 average comments, we have to tell them 'bad', 'better' and 'good' somehow...

...unless you are thinking that literally just a bar 

xxxxxxxxxxxxxxx--------------20

x-----------------------------243 

would accomplish this. I'm not sure. Without zones as shown in my image above, their progress shrinks as per Mike's point (shown with the xxxxxxx becomeing x as more comments are submitted). I think I would hate that. The zones give you a target even if there are 600 comments (this happens)

**Comment by patcon** [comment]
haha fair enough. I'll find omnigraffle or something

**Comment by colinmegill** [comment]
We use Sketch 3 and it is amazing.

**Comment by patcon** [comment]
hey, just winging it with inkscape, but i'll investigate that later :)

![flat progress bar](https://imgur.com/sYOlxzb.png)

basically, i simply disagree that the role of this progress bar should be to inform users that "there will be more comments and you should come back". So I respectfully disagree that attempting to colour them on that premise is appropriate (at least not at this point in a possible march toward game-fication). That idea of returning to the poll needs to be communicated, but this colour can't seem to do that subtext justice. Users just want to know how many are left, not whether answering less than 5 is "bad" (RED! WARNING FOR SOME REASON EVEN THOUGH YOU MAYBE CAN'T RESOLVE IT.) or that 20 is good. It's noise that doesn't address what a user is wondering in its absence imho.

Of course, I am just one opinion. And I totes understand if you disagree.

**Comment by colinmegill** [comment]
That's fair, the bar does not and should not communicate 'coming back.' I agree. When participants finish a small number of comments, there is already copy to handle 'coming back' connected to the email form at the last comment.

But I do think it should communicate some reasonable target, in the case that there are 600 comments or more. Here are a couple ideas from my side.

Total progress, target progress, total comments, all on one compact line:

![image](https://cloud.githubusercontent.com/assets/1770265/21207631/d884ab64-c22e-11e6-869b-6ebc408fbbe4.png)

When total comment count is below target vote count, the thick bar is pinned to the end: 

![image](https://cloud.githubusercontent.com/assets/1770265/21207190/4de280aa-c22c-11e6-9ef5-811d2716194b.png)

**Comment by patcon** [comment]
What if there was absolutely no threshold number, but we just show what percentile people are in number of responses? Or as I'm sure you'd prefer, communicate it visually ;)

**Comment by colinmegill** [comment]
Oh, quintiles? So make them compete?

**Comment by colinmegill** [comment]
Hadn't considered that approach, that's interesting. Send some mocks if you like.

**Comment by colinmegill** [comment]
Just so you know that I'm hearing you, you want this:

![image](https://cloud.githubusercontent.com/assets/1770265/21213516/7cbfec36-c25a-11e6-9f12-ec31a0b44b0d.png)

I've never seen a progress bar that is not supposed to finish. I would expect that this strongly suggests to the user that they are expected to finish.

This returns to the very first line of the issue - what it means to be 'done'. Currently we have nothing explicit because of exactly this reason. We've avoided adding anything because this really changes over different scales. "Only 9,252 left to go!" Will be quite comical.

**Comment by colinmegill** [comment]
@patcon I think I have a solution here. We'll go with the basic progress bar. If the number is over 100 comments, we'll say "Whoa! There are a lot of comments. 100 is a good amount to vote on." Underneath of it, or something like that. The progress bar can stand alone, as you suggest, and we'll handle the edge cases with language specific to the edge cases.

**Comment by patcon** [comment]
That's definitely a improvement!

having said that, maybe my progress bar ask was a distraction on my part. fwiw, for myself, i kinda just want to know how many cards are in the system. I wasn't trying to gamify it any more than a subway sign that says "3 minutes till next train" is trying to gamify subways :) As an admin, I get to see the number of comments in the backend, but that's not transparent to a regular user. As an engaged user, I think I'd want to know the number without spin, just so I can see inside this thing a bit -- maybe see the number today, and the number tomorrow, a and decide how it will affect my engagement, and maybe I'll do a little informal "slope" calculation in my head, or whatever :)

In my thinking, something like this is just as helpful as a progress bar, and perhaps more clear in meaning in some ways (assuming I'm watching the numbers change as I click through and cards change stacks):
![numbered card stack icon suggestion](https://imgur.com/cXv87PU.png)

(unseen card -> voted cards)

**Comment by colinmegill** [comment]
Well, now we're talking. 

![image](https://cloud.githubusercontent.com/assets/1770265/22756092/da93a1f6-edfa-11e6-888d-777c5d86249b.png)

**Comment by colinmegill** [comment]
There was one more iteration, and it's live! 

![image](https://user-images.githubusercontent.com/1770265/28735057-7971ba9c-7398-11e7-9ff2-2c251a46d125.png)

**Comment by colinmegill** [comment]
As a final thought / check in here it would be useful to see later in an A/B test the influence of showing this only after 10 votes, to get people invested in voting before they realize how many there are, and also, secondarily, to let them know they don't have to vote on all of them.
